File: VB/E965.Module/Controllers/TestListViewController.vb
Imports System
Imports System.IO
Imports System.Linq
Imports FileSystemData
Imports DevExpress.ExpressApp
Imports System.Threading.Tasks
Imports DevExpress.Persistent.Base
Imports System.Collections.Generic
Imports DevExpress.ExpressApp.Actions
Imports FileSystemData.BusinessObjects
Imports DevExpress.Persistent.Validation

Namespace E965.Module.Controllers
	Partial Public Class TestListViewController
		Inherits ViewController

		Private Const MaxTestersCount As Integer = 10
		Private Const STR_Testfile As String = "Test.file"
		Private testAction As SimpleAction
		Public Sub New()
			testAction = New SimpleAction(Me, "Test", "View")
			TargetViewType = ViewType.ListView
			AddHandler testAction.Execute, AddressOf testAction_Execute
		End Sub
		Private Sub testAction_Execute(ByVal sender As Object, ByVal e As SimpleActionExecuteEventArgs)
			testAction.Caption = "Test is running..."
			Dim tasks(MaxTestersCount - 1) As Task
			For i As Integer = 0 To MaxTestersCount - 1
				tasks(i) = Task.Factory.StartNew(Sub()
					For j As Integer = 0 To MaxTestersCount - 1
						Dim testStream As Stream = Me.GetType().Assembly.GetManifestResourceStream("Test.file")
						Using os As IObjectSpace = Application.CreateObjectSpace()
							Try
								Dim theObject As IFileData = os.CreateObject(Of FileSystemStoreObject)()
								If Not DirectCast(theObject, IEmptyCheckable).IsEmpty Then
									Throw New Exception("Must be empty")
								End If
								theObject.LoadFromStream(STR_Testfile, testStream)
								If theObject.FileName <> STR_Testfile Then
									Throw New Exception("Must have file name")
								End If
								os.CommitChanges()
								If theObject.Size <> 44 Then
									Throw New Exception("Must have size 44")
								End If
								If DirectCast(theObject, IEmptyCheckable).IsEmpty Then
									Throw New Exception("Must be non-empty")
								End If
							Catch
								os.Rollback()
								Throw
							End Try
						End Using
					Next j
				End Sub)
			Next i
			Try
				Task.WaitAll(tasks)
				View.ObjectSpace.Refresh()
				testAction.Caption = "Succeeded"
			Catch ex As Exception
				testAction.Caption = "Failed"
				Tracing.Tracer.LogError(ex)
			End Try
			testAction.Enabled("Run only once") = False
		End Sub
	End Class
End Namespace

File: CS/E965.Module/Controllers/TestListViewController.cs
using System;
using System.IO;
using System.Linq;
using FileSystemData;
using DevExpress.ExpressApp;
using System.Threading.Tasks;
using DevExpress.Persistent.Base;
using System.Collections.Generic;
using DevExpress.ExpressApp.Actions;
using FileSystemData.BusinessObjects;
using DevExpress.Persistent.Validation;

namespace E965.Module.Controllers {
    public partial class TestListViewController : ViewController {
        private const int MaxTestersCount = 10;
        private const string STR_Testfile = "Test.file";
        private SimpleAction testAction;
        public TestListViewController() {
            testAction = new SimpleAction(this, "Test", "View");
            TargetViewType = ViewType.ListView;
            testAction.Execute += testAction_Execute;
        }
        void testAction_Execute(object sender, SimpleActionExecuteEventArgs e) {
            testAction.Caption = "Test is running...";
            Task[] tasks = new Task[MaxTestersCount];
            for (int i = 0; i < MaxTestersCount; i++) {
                tasks[i] = Task.Factory.StartNew(() => {
                    for (int j = 0; j < MaxTestersCount; j++) {
                        Stream testStream = GetType().Assembly.GetManifestResourceStream("E965.Module.Controllers.Test.file");
                        using (IObjectSpace os = Application.CreateObjectSpace(typeof(FileSystemStoreObject))) {
                            try {
                                IFileData theObject = os.CreateObject<FileSystemStoreObject>();
                                if (!((IEmptyCheckable)theObject).IsEmpty)
                                    throw new Exception("Must be empty");
                                theObject.LoadFromStream(STR_Testfile, testStream);
                                if (theObject.FileName != STR_Testfile)
                                    throw new Exception("Must have file name");
                                os.CommitChanges();
                                if (theObject.Size != 44)
                                    throw new Exception("Must have size 44");
                                if (((IEmptyCheckable)theObject).IsEmpty)
                                    throw new Exception("Must be non-empty");

                            } catch {
                                os.Rollback();
                                throw;
                            }
                        }
                    }
                }
                );
            }
            try {
                Task.WaitAll(tasks);
                View.ObjectSpace.Refresh();
                testAction.Caption = "Succeeded";
            } catch (Exception ex) {
                testAction.Caption = "Failed";
                Tracing.Tracer.LogError(ex);
            }
            testAction.Enabled["Run only once"] = false;
        }
    }
}